src/NumericalAlgorithms/RootFinding/GslMultiRootFunctionFdf.cpp

```diff
@@ -0,0 +1,76 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "NumericalAlgorithms/RootFinding/GslMultiRootFunctionFdf.hpp"
+
+#include <array>
+#include <gsl/gsl_errno.h>
+#include "Utilities/Gsl.hpp"
+
+template <size_t Dim>
+int gsl_multirootfunctionfdf_wrapper_f(const gsl_vector* const v, void* params,
+                                       gsl_vector* const func) noexcept {
+  const GslMultiRootFunctionFdf<Dim>* fn =
+      static_cast<const GslMultiRootFunctionFdf<Dim>*>(params);
+  std::array<double, Dim> input_as_std_array{};
+  for (size_t i = 0; i < Dim; i++) {
+    gsl::at(input_as_std_array, i) = (gsl_vector_get(v, i));
+  }
+  const std::array<double, Dim> result_as_std_array = fn->f(input_as_std_array);
+  for (size_t i = 0; i < Dim; i++) {
+    gsl_vector_set(func, i, gsl::at(result_as_std_array, i));
+  }
+  return GSL_SUCCESS;
+}
+
+template <size_t Dim>
+int gsl_multirootfunctionfdf_wrapper_df(const gsl_vector* const v, void* params,
+                                        gsl_matrix* const J) noexcept {
+  const GslMultiRootFunctionFdf<Dim>* fn =
+      static_cast<const GslMultiRootFunctionFdf<Dim>*>(params);
+  std::array<double, Dim> input_as_std_array{};
+  for (size_t i = 0; i < Dim; i++) {
+    gsl::at(input_as_std_array, i) = (gsl_vector_get(v, i));
+  }
+  const std::array<std::array<double, Dim>, Dim> jacobian =
+      fn->df(input_as_std_array);
+  for (size_t i = 0; i < Dim; i++) {
+    for (size_t j = 0; j < Dim; j++) {
+      gsl_matrix_set(J, i, j, gsl::at(gsl::at(jacobian, i), j));
+    }
+  }
+  return GSL_SUCCESS;
+}
+
+template <size_t Dim>
+int gsl_multirootfunctionfdf_wrapper_fdf(const gsl_vector* const v,
+                                         void* params, gsl_vector* const func,
+                                         gsl_matrix* J) noexcept {
```

Comment:
Mark `params` and `J` const.  (Also `params` above.)
J -> jacobian

Reply:
I think gsl only accepts the function if it's a void* - it didn't complain about the other ones being const though, but I can check again